src/org/thoughtcrime/securesms/loki/protocol/ClosedGroupsProtocolV2.kt

```diff
@@ -282,7 +444,134 @@ object ClosedGroupsProtocolV2 {
         LokiPushNotificationManager.performOperation(context, ClosedGroupOperation.Subscribe, groupPublicKey, userPublicKey)
     }
 
-    public fun handleClosedGroupUpdate(context: Context, closedGroupUpdate: SignalServiceProtos.ClosedGroupUpdateV2, sentTimestamp: Long, groupPublicKey: String, senderPublicKey: String) {
+    fun handleClosedGroupMembersRemoved(context: Context, closedGroupUpdate: SignalServiceProtos.ClosedGroupUpdateV2, sentTimestamp: Long, groupPublicKey: String, senderPublicKey: String) {
+        val apiDB = DatabaseFactory.getLokiAPIDatabase(context)
+        val groupDB = DatabaseFactory.getGroupDatabase(context)
+        val groupID = doubleEncodeGroupID(groupPublicKey)
+        val group = groupDB.getGroup(groupID).orNull()
+        if (group == null) {
+            Log.d("Loki", "Ignoring closed group info message for nonexistent group.")
+            return
+        }
+        val userPublicKey = TextSecurePreferences.getLocalNumber(context)
+        val name = group.title
+        // Check common group update logic
+        val members = group.members.map { it.serialize() }
+        val admins = group.admins.map { it.toPhoneString() }
+
+        // Users that are part of this remove update
+        val updateMembers = closedGroupUpdate.membersList.map { it.toByteArray().toHexString() }
+
+        if (!isValidGroupUpdate(group, sentTimestamp, senderPublicKey)) {
+            return
+        }
+        // If admin leaves the group is disbanded
+        val didAdminLeave = admins.any { it in updateMembers }
+        // newMembers to save is old members minus removed members
+        val newMembers = members - updateMembers
+        val senderLeft = senderPublicKey in updateMembers
```

Comment:
Under normal circumstances the user should send a `MEMBER_LEFT` when they leave and not a `MEMBERS_REMOVED`, but I guess it doesn't hurt to check :)